apps/email/js/sync.js

```diff
@@ -336,28 +345,40 @@ define(function(require) {
             };
 
             if (model.getAccountCount() === 1) {
-              subject = info.subject;
-              body = info.from;
+              subjectL10n = { raw: info.subject };
+              bodyL10n = { raw: info.from };
             } else {
-              subject = mozL10n.get('new-emails-notify-multiple-accounts', {
-                n: count,
-                accountName: result.address
-              });
-              body = mozL10n.get('new-emails-notify-multiple-accounts-body', {
-                from: info.from,
-                subject: info.subject
-              });
+              subjectL10n = {
+                id: 'new-emails-notify-multiple-accounts',
+                args: {
+                  n: count,
+                  accountName: result.address
+                }
+              };
+              bodyL10n = {
+                id: 'new-emails-notify-multiple-accounts-body', 
+                args: {
+                  from: info.from,
+                  subject: info.subject
+                }
+              };
             }
           }
 
-          sendNotification(
-            result.id,
-            subject,
-            body,
-            iconUrl,
-            dataObject,
-            behavior
-          );
+          if (!bodyL10n.then) {
+            bodyL10n = Promise.resolve(bodyL10n);
+          }
+          
+          bodyL10n.then(result => {
+            sendNotification(
+              result.id,
+              subjectL10n,
+              result,
+              iconUrl,
+              dataObject,
+              behavior
+            );
+          });
```

Comment:
This bit here is a bit tricky since the Promise then resolution will be an async resolution, after the accountsResults.updates.forEach() iteration has finished, and after the finishSync() after the accountsResults.updates.forEach() has been called.

We want to only call finishSync when we know all the work of sending out the notifications has been done.

Since the issue is that we really need to wait on `mozL10n.formatValue('senders-separation-sign')`'s promise to resolve, to avoid too much nesting of async blocks, let's move the `mozL10n.formatValue('senders-separation-sign')` work into the `getEnv function. It can be the innermost step of getEnv(). This also means the fn that getEnv calls will pass the `separator` variable as a fourth argument.

This this means the `if (!bodyL10n.then) {` block is not needed, and there is no more need for the `bodyL10n.then(result => {` part, the sendNotification can be called directly.